etcdserver/api/v3rpc/watch.go

```diff
@@ -117,10 +123,33 @@ func (sws *serverWatchStream) recvLoop() error {
 				id = sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
 			}
 			sws.ctrlStream <- &pb.WatchResponse{
-				Header:   sws.newResponseHeader(wsrev),
-				WatchId:  int64(id),
-				Created:  true,
-				Canceled: futureRev,
+				Header:        sws.newResponseHeader(wsrev),
+				WatchId:       int64(id),
+				Created:       true,
+				Canceled:      futureRev,
+				StartRevision: rev,
+			}
+
+			if uv.CreateRequest.ProgressReport {
+				go func() {
+					for {
+						select {
+						case <-sws.reportDone:
+							close(sws.reportStream)
+							return
+						case <-time.After(time.Duration(uv.CreateRequest.ReportInterval) * time.Second):
+							if sws.watchStream.WatchStatus(int64(id)) != storage.Synced {
```

Comment:
@xiang90 Just added this method from `storage` package. Can you take a look? Thanks.
(I still have to figure out the fix for race conditions that Anthony pointed out).